### PR TITLE
Bigquery: add table schemas

### PR DESCRIPTION
**Pull request by tseaver**
Based on #1021

### CODE REVIEW

**Comment by dhermes** [comment]
Ping me when #1021 is in. We are pretty close to having that one in AFAICT.

**Comment by tseaver** [comment]
@dhermes rebased after merging #1021.

File: gcloud/bigquery/table.py [hunk 1]
```diff
@@ -22,6 +22,35 @@
...
 
+class SchemaField(object):
+    """Describe a single field within a table schema.
+
+    :type name: string
+    :param name: the name of the field
+
+    :type field_type: string
+    :param field_type: the type of the field (one of 'STRING', 'INTEGER',
+                       'FLOAT', 'BOOLEAN', 'TIMESTAMP' or 'RECORD')
```
Comment (dhermes) [inline gcloud/bigquery/table.py]: Do we want to start using https://pypi.python.org/pypi/enum?

File: gcloud/bigquery/table.py [hunk 2]
```diff
@@ -22,6 +22,35 @@
...
+
+    :type mode: string
+    :param mode: the type of the field (one of 'NULLABLE', 'REQUIRED',
+                 or 'REPEATED')
+
+    :type description: string
+    :param description: optional description for the field
+
+    :type fields: list of ``SchemaField``, or None
```
Comment (dhermes) [inline gcloud/bigquery/table.py]:
```
:class:`SchemaField`
```

will be more useful with Sphinx

File: gcloud/bigquery/table.py [hunk 3]
```diff
@@ -50,6 +83,32 @@ def path(self):
...
+        :raises: TypeError if 'value' is not a sequence, or ValueError if
+                 any item in the sequence is not a SchemaField
+        """
+        fields = list(value)
+        if len(fields) > 0:
+            types = set([type(field) for field in fields])
+            if types != set([SchemaField]):
+                raise ValueError('Schema items must be fields')
+        self._schema = tuple(value)
+
```
Comment (dhermes) [inline gcloud/bigquery/table.py]: Worth having a way to append to `_schema`?

File: gcloud/bigquery/table.py [hunk 4]
```diff
@@ -50,6 +83,32 @@ def path(self):
...
+        :type value: list of ``SchemaField``
+        :param value: fields describing the schema
+
+        :raises: TypeError if 'value' is not a sequence, or ValueError if
+                 any item in the sequence is not a SchemaField
+        """
+        fields = list(value)
+        if len(fields) > 0:
+            types = set([type(field) for field in fields])
```
Comment (dhermes) [inline gcloud/bigquery/table.py]: No need to create a new list or a set:

```python
all(isinstance(field, SchemaField) for field in value)
```

Probably irrelevant given the tiny sizes of inputs.

File: gcloud/bigquery/table.py [hunk 5]
```diff
@@ -50,6 +83,32 @@ def path(self):
...
+            if types != set([SchemaField]):
+                raise ValueError('Schema items must be fields')
+        self._schema = tuple(value)
```
Comment (dhermes) [inline gcloud/bigquery/table.py]: Why cast to `list` at beginning and then to `tuple` here? (I understand the tuple cast, just not the first one.)

File: gcloud/bigquery/test__helpers.py [hunk 6]
```diff
@@ -41,7 +41,8 @@ def test_w_millis(self):
         import datetime
         import pytz
         from gcloud.bigquery._helpers import _total_seconds
-        NOW = datetime.datetime.utcnow().replace(tzinfo=pytz.utc)
+        NOW = datetime.datetime(2015, 7, 29, 17, 45, 21, 123456,
+                                tzinfo=pytz.utc)
```
Comment (dhermes) [inline gcloud/bigquery/test__helpers.py]: Seems like a good choice for tests. Was the `utcnow()` causing issues or you just wanted a deterministic test?

File: gcloud/bigquery/table.py (hunk 1 above)
Comment (tseaver) [inline gcloud/bigquery/table.py]: I don't know.  If we do pick one, then I think the PEP 435-compatible https://pypi.python.org/pypi/enum34/ is probably the way to go:  it ports the PEP, which landed in Python 3.4, back to all the versions we care about (and more).

File: gcloud/bigquery/test__helpers.py (hunk 6 above)
Comment (tseaver) [inline gcloud/bigquery/test__helpers.py]: Both (rounding errors suck).

File: gcloud/bigquery/table.py (hunk 5 above)
Comment (tseaver) [inline gcloud/bigquery/table.py]: It was an "iterability" assertion, now gone in a4e78d5

File: gcloud/bigquery/table.py (hunk 4 above)
Comment (tseaver) [inline gcloud/bigquery/table.py]: a4e78d5

File: gcloud/bigquery/table.py (hunk 3 above)
Comment (tseaver) [inline gcloud/bigquery/table.py]: I felt like having it be assignable-but-immutable made more sense (modifying it after the table is created on the back-end is particularly restricted, which I don't try to capture here).

File: gcloud/bigquery/table.py (hunk 2 above)
Comment (tseaver) [inline gcloud/bigquery/table.py]: 938ab1b

File: gcloud/bigquery/table.py (hunk 1 above)
Comment (tseaver) [inline gcloud/bigquery/table.py]: Opened https://github.com/GoogleCloudPlatform/gcloud-python/issues/1031 to track this issue (which affects more places than this).

File: gcloud/bigquery/table.py [hunk 7]
```diff
@@ -50,6 +83,29 @@ def path(self):
...
+
+        :type value: list of :class:`SchemaField`
+        :param value: fields describing the schema
+
+        :raises: TypeError if 'value' is not a sequence, or ValueError if
+                 any item in the sequence is not a SchemaField
+        """
+        if not all(isinstance(field, SchemaField) for field in value):
+            raise ValueError('Schema items must be fields')
+        self._schema = tuple(value)
```
Comment (dhermes) [inline gcloud/bigquery/table.py]: > I felt like having it be assignable-but-immutable made more sense (modifying it after the table is created on the back-end is particularly restricted, which I don't try to capture here).

You mean once an `insert` has occurred, `update` won't allow updating schema?

**Comment by dhermes** [comment]
LGTM.

I don't want the `schema.setter` to be a blocker, but do want to understand it.

File: gcloud/bigquery/table.py (hunk 7 above)
Comment (tseaver) [inline gcloud/bigquery/table.py]: Only certain kinds of changes are allowed.   See: https://cloud.google.com/bigquery/docs/tables#updateschema

**Pull request merged**